contrib/format-excel/src/main/java/org/apache/drill/exec/store/excel/ExcelBatchReader.java

```diff
@@ -0,0 +1,398 @@
+              //.replaceAll("_", "__")  // Not sure if we need this
+              .replace(PARSER_WILDCARD, SAFE_WILDCARD).replaceAll("\\.", SAFE_SEPARATOR);
+            ExcelFormatPlugin.makeColumn(builder, tempColumnName, TypeProtos.MinorType.VARCHAR);
+            excelFieldNames.add(colPosition, tempColumnName);
+            break;
+          case NUMERIC:
+            tempColumnName = String.valueOf(cell.getNumericCellValue());
+            ExcelFormatPlugin.makeColumn(builder, tempColumnName, TypeProtos.MinorType.FLOAT8);
+            excelFieldNames.add(colPosition, tempColumnName);
+            break;
+        }
+        colPosition++;
+      }
+    }
+    return builder.buildSchema();
+  }
+
+  /**
+   * Helper function to get the selected sheet from the configuration
+   *
+   * @return XSSFSheet The selected sheet
+   */
+  private XSSFSheet getSheet() {
+    int sheetIndex = 0;
+    if (!readerConfig.sheetName.isEmpty()) {
+      sheetIndex = workbook.getSheetIndex(readerConfig.sheetName);
+    }
+
+    //If the sheet name is not valid, throw user exception
+    if (sheetIndex == -1) {
+      throw UserException.dataReadError().message("Could not open sheet " + readerConfig.sheetName).build(logger);
+    } else {
+      return workbook.getSheetAt(sheetIndex);
+    }
+  }
+
+  @Override
+  public boolean next() {
+    RowSetLoader rowWriter = loader.writer();
+    recordCount = 0;
+    while (!rowWriter.isFull()) {
+      if (!nextLine(rowWriter)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public boolean nextLine(RowSetLoader rowWriter) {
+    if (!rowIterator.hasNext()) {
+      return false;
+    } else if (recordCount >= readerConfig.lastRow) {
+      return false;
+    }
+
+    int lastRow = readerConfig.lastRow;
+    while (recordCount < lastRow && rowIterator.hasNext()) {
+
+      lineCount++;
+
+      Row row = rowIterator.next();
+      // If the user specified that there are no headers, get the column count
+      if (readerConfig.headerRow == -1 && recordCount == 0) {
+        this.totalColumnCount = row.getLastCellNum();
+      }
+
+      String fieldName;
+      if (row.getLastCellNum() < totalColumnCount) {
+        throw UserException.dataReadError().message("Wrong number of columns in row: %d", row.getLastCellNum()).build(logger);
+      }
+      int colPosition = 0;
+      if (readerConfig.firstColumn != 0) {
+        colPosition = readerConfig.firstColumn - 1;
+      }
+
+      int finalColumn = totalColumnCount;
+      if (readerConfig.lastColumn != 0) {
+        finalColumn = readerConfig.lastColumn - 1;
+      }
+      rowWriter.start();
+      for (; colPosition < finalColumn; colPosition++) {
+        Cell cell = row.getCell(colPosition);
+
+        CellValue cellValue = evaluator.evaluate(cell);
+        fieldName = excelFieldNames.get(colPosition);
+
+        // Remove newlines from field names
+        fieldName = fieldName.replaceAll("\\n", SAFE_NEWLINE_REPLACEMENT);
```

Comment:
Similar to a comment made in another PR... We have a fixed set of columns in a spreadsheet. Column names must, presumably, be the same for a column in each row in which it appears:

|   | A | B |
 -- | -- | -- |
| 1 | name | balance |
| 2 | fred | 123.45 |
| 3 | barney | 98.76 |

Given this, you can create an array of `ScalarWriter` objects, then index into this array as you read each column. This saves doing he column name edits, and name lookup, on every cell.